### PR TITLE
[REEF-1658] Gracefully shut down all threads at the end of the REEF job

### PR DESCRIPTION
**Pull request by motus**
This work is part of "REEF as a library" project [REEF-1561](https://issues.apache.org/jira/browse/REEF-1561)

After this update, REEF Driver should be able to close *without* explicit `System.exit(0)` invocation at the end. We will keep that call until we test all corner cases more thoroughly to make sure we always clean up the resources on REEF job exit.

Summary of changes:
  * Close the `EvaluatorIdlenessThreadPool` in `DriverRuntimeStopHandler`
  * Make `EvaluatorIdlenessThreadPool` `AutoCloseable` to close the local thread pool
  * Implement `EvaluatorManager.shutdown()` method and call it when removing evaluator from `Evaluators`
  * Make sure `.close()` methods never throw in `DispatchingEStage` and `EvaluatorMessageDispatcher`
  * Give better thread names in `EvaluatorIdlenessThreadPool`
  * Improve logging and error handling in `.close()` methods in all path of the driver shutdown

JIRA: [REEF-1658](https://issues.apache.org/jira/browse/REEF-1658)

### CODE REVIEW

**Comment by tcNickolas** [comment]
Just making sure that you know that this change fails Java build.

**Comment by markusweimer** [comment]
> Just making sure that you know that this change fails Java build.

Thanks, @tcNickolas! @motus, is this because of the need to have #1174 merged before?

**Comment by motus** [comment]
@markusweimer @tcNickolas exactly. It depends on #1174 (assuming that certain `.close()` calls don't throw)

**Comment by markusweimer** [comment]
@motus Can you please rebase before we start reviewing this? There is now a merge from `master` in this PR, which is always tricky when it comes to merging this.

**Comment by motus** [comment]
@markusweimer I am done with the rebase. I am re-running tests locally, just in case. Meanwhile, you can go ahead and review the code.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorIdlenessThreadPool.java [hunk 1]
```diff
@@ -60,21 +65,60 @@ private EvaluatorIdlenessThreadPool(@Parameter(EvaluatorIdlenessThreadPoolSize.c
...
+    try {
+      isTerminated = this.executor.awaitTermination(this.waitInMillis, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException ex) {
+      LOG.log(Level.WARNING, "EvaluatorIdlenessThreadPool shutdown: Interrupted", ex);
+    }
+
+    if (isTerminated) {
+      LOG.log(Level.FINE, "EvaluatorIdlenessThreadPool shutdown: Terminated successfully");
+    } else {
+      final List<Runnable> pendingJobs = this.executor.shutdownNow();
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorIdlenessThreadPool.java]: We might want to log that we got here. This should usually not happen, right?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java [hunk 2]
```diff
@@ -219,6 +219,8 @@ public EvaluatorDescriptor getEvaluatorDescriptor() {
   @Override
   public void close() {
 
+    LOG.log(Level.FINER, "Closing: {0}", this.toString());
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java]: `toString` might be too big. Maybe log `Closing EvaluatorManager for Evaluator "{evaluatorId}"`

**Comment by motus** [comment]
Done with the fixes! @markusweimer please have a look. Thanks!

**Comment by markusweimer** [comment]
LGTM, will test & merge

**Comment by markusweimer** [comment]
When running tests on Java, I get many instances of the following exception:

```
SEVERE: Error closing the acceptor channel for /192.168.1.2:17132
java.util.concurrent.RejectedExecutionException: event executor terminated
        at io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:735)
        at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:312)
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:718)
        at io.netty.channel.AbstractChannelHandlerContext.safeExecute(AbstractChannelHandlerContext.java:884)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:555)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:419)
        at io.netty.channel.DefaultChannelPipeline.close(DefaultChannelPipeline.java:828)
        at io.netty.channel.AbstractChannel.close(AbstractChannel.java:179)
        at org.apache.reef.wake.remote.transport.netty.NettyMessagingTransport.close(NettyMessagingTransport.java:224)
        at org.apache.reef.io.network.naming.NameClient.close(NameClient.java:154)
        at org.apache.reef.io.network.util.NetworkMessagingTestService.close(NetworkMessagingTestService.java:104)
        at org.apache.reef.io.network.NetworkConnectionServiceTest.testMessagingNetworkConnServiceRate(NetworkConnection
ServiceTest.java:214)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Nov 02, 2016 11:18:03 AM org.apache.reef.wake.remote.transport.netty.NettyMessagingTransport close
SEVERE: Error closing the acceptor channel for /192.168.1.2:16400
java.util.concurrent.RejectedExecutionException: event executor terminated
        at io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:735)
        at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:312)
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:718)
        at io.netty.channel.AbstractChannelHandlerContext.safeExecute(AbstractChannelHandlerContext.java:884)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:555)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:419)
        at io.netty.channel.DefaultChannelPipeline.close(DefaultChannelPipeline.java:828)
        at io.netty.channel.AbstractChannel.close(AbstractChannel.java:179)
        at org.apache.reef.wake.remote.transport.netty.NettyMessagingTransport.close(NettyMessagingTransport.java:224)
        at org.apache.reef.io.network.naming.NameClient.close(NameClient.java:154)
        at org.apache.reef.io.network.util.NetworkMessagingTestService.close(NetworkMessagingTestService.java:105)
        at org.apache.reef.io.network.NetworkConnectionServiceTest.testMessagingNetworkConnServiceRate(NetworkConnection
ServiceTest.java:214)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```

Is this expected?

**Comment by motus** [comment]
@markusweimer it is expected - we are trying to close the acceptor that has already been closed, hence the error. Maybe we can downgrade it to a warning again? :smiley:

Anyway, I was meant to open another JIRA to describe that condition to double check if we close the same transport more than once. Meanwhile, I think the most important thing is to keep the `.close()` method idempotent.

**Comment by markusweimer** [comment]
@motus Fair enough. Tests still pass, so I will go ahead with the merge.

**Comment by motus** [comment]
Created a JIRA [REEF-1661](https://issues.apache.org/jira/browse/REEF-1661) to track the issue with `RejectedExecutionException`
